### PR TITLE
feat: support reloading cloud controller manager from secret dynamically

### LINKED ISSUE #604: feature: dynamically reloading configurations from secret
**Is your feature request related to a problem? Please describe.**

The configuration of cloud provider config file may be changed later after deployment; however, the new configurations won't take effect until the cloud controller manager is restarted. Since the azure cloud provider has already supported [initializing itself from a secret](https://github.com/kubernetes-sigs/cloud-provider-azure/blob/91134fdb2549a36144ca35b97f544bf075e9e38b/pkg/provider/azure_config.go#L43-L58) , it could also re-initialize itself via watching the secret changes, and then won't need another restarting. 

**Describe the solution you'd like in detail**

- Remove requirement of cloud config file when initializing from secret
- Watch changes from secret and then re-initialize cloud provider if secret has changed

**Describe alternatives you've considered**

Restart cloud controller manager manually after config changes before this new feature is supported.

**Additional context**

### PR DESCRIPTION
**Pull request by nilo19**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

/kind feature

**What this PR does / why we need it**:

This PR enables reloading ccm from secret dynamically.
There are two ways to build a ccm: statically build from cloud-config file or dynamically build from secret.
To use the cloud-config file, set `--enable-dynamic-reloading=false` and `--cloud-config` to an non-empty value. To use the secret, set `--enable-dynamic-reloading=true`, `--cloud-config-secret-name`, `--cloud-config-secret-namespace` and `--cloud-config-key`. In this way, the ccm would be restarted every time a change is made to the secret.

Works to do:
- [ ] update the doc

**Which issue(s) this PR fixes**:

Fixes #604

**Special notes for your reviewer**:

**Release note**:
```
feat: support reloading cloud controller manager from secret dynamically
```

### CODE REVIEW

File: cmd/cloud-controller-manager/app/options/dynamic.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+}
+
+// AddFlags adds flags releated to dynamic reloading for controller manager to the specified FlagSet
+func (o *DynamicReloadingOptions) AddFlags(fs *pflag.FlagSet) {
+	if o == nil {
+		return
+	}
+
+	fs.BoolVar(&o.EnableDynamicReloading, "enable-dynamic-reloading", false, "Enable re-configuring cloud controller manager from secret without restarting")
+	fs.StringVar(&o.CloudConfigSecretName, "cloud-config-secret-name", "cloud-provider-config", "The name of the cloud config secret, default to 'cloud-config'")
```
Comment (feiskyer) [inline cmd/cloud-controller-manager/app/options/dynamic.go:40]: help message is inconsistent with default value 'cloud-provider-config'

File: cmd/cloud-controller-manager/app/options/dynamic.go [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+	fs.StringVar(&o.CloudConfigSecretNamespace, "cloud-config-secret-namespace", "kube-system", "The k8s namespace of the cloud config secret, default to 'kube-system'")
+	fs.StringVar(&o.CloudConfigKey, "cloud-config-key", "cloud-config", "The key of the config data in the cloud config secret, default to 'azure-cloud-provider'")
```
Comment (feiskyer) [inline cmd/cloud-controller-manager/app/options/dynamic.go:42]: default to 'cloud-config'

File: cmd/cloud-controller-manager/app/dynamic/watch_secret.go [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+	"fmt"
+	"os"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/util/wait"
+	cloudcontrollerconfig "sigs.k8s.io/cloud-provider-azure/cmd/cloud-controller-manager/app/config"
+	"sigs.k8s.io/cloud-provider-azure/cmd/cloud-controller-manager/app/options"
+
+	"k8s.io/klog/v2"
+
```
Comment (feiskyer) [inline cmd/cloud-controller-manager/app/dynamic/watch_secret.go:29]: nit: remove empty line

File: cmd/cloud-controller-manager/app/controllermanager.go [hunk 4]
```diff
@@ -255,9 +334,13 @@ func startControllers(c *cloudcontrollerconfig.CompletedConfig, stopCh <-chan st
 		klog.Fatalf("Failed to wait for apiserver being healthy: %v", err)
 	}
 
+	klog.V(2).Infof("startControllers: starting shared informers")
 	c.SharedInformers.Start(stopCh)
 
-	select {}
+	<-stopCh
+	klog.V(1).Infof("startControllers: received stopping signal, returning")
```
Comment (feiskyer) [inline cmd/cloud-controller-manager/app/controllermanager.go:341]: s/returning/exiting

**Review by feiskyer — COMMENTED** [review]
/lgtm
/approve

**Comment by andyzhangx** [comment]
shall we set `az.SecretNamespace` as `kube-system` and `az.SecretName` as `azure-cloud-provider` by default? currently it's empty, not consistent with doc: https://github.com/kubernetes-sigs/azuredisk-csi-driver/blob/master/docs/read-from-secret.md, and user would hit following error when read cloud config from secret:
```
I0610 14:51:11.909193       1 azure.go:56] reading cloud config from secret
I0610 14:51:11.909246       1 request.go:857] Error in request: resource name may not be empty
E0610 14:51:11.909292       1 azure_config.go:45] Failed to get cloud-config from secret: failed to get secret /: resource name may not be empty
I0610 14:51:11.909307       1 azure.go:59] InitializeCloudFromSecret failed with error: InitializeCloudFromSecret: failed to get cloud config from secret /: failed to get secret /: resource name may not be empty
I0610 14:51:11.909311       1 azure.go:64] could not read cloud config from secret
I0610 14:51:11.909329       1 azure.go:75] use default AZURE_CREDENTIAL_FILE env var: /etc/kubernetes/azure.json
``` 

Should add a test to prevent such regression.

**Comment by feiskyer** [comment]
> shall we set az.SecretNamespace as kube-system and az.SecretName as azure-cloud-provider by default? 

makes sense to me. @nilo19 let's change the default configuration back.

**Comment by nilo19** [comment]
https://github.com/kubernetes-sigs/cloud-provider-azure/blob/master/cmd/cloud-controller-manager/app/options/dynamic.go#L41

Not sure why the default value defined here didn't work as expected.

**Comment by andyzhangx** [comment]
> https://github.com/kubernetes-sigs/cloud-provider-azure/blob/master/cmd/cloud-controller-manager/app/options/dynamic.go#L41
> 
> Not sure why the default value defined here didn't work as expected.

@nilo19 default value only works for ccm, cloud provider needs the default value too.

**Comment by nilo19** [comment]
@andyzhangx the cloud provider receives the secret name & namespace from the ccm configs.

https://github.com/kubernetes-sigs/cloud-provider-azure/blob/47def28466a419b5d917efadafcf48e33f0880e3/cmd/cloud-controller-manager/app/controllermanager.go#L264

**Comment by andyzhangx** [comment]
csi driver depends on cloud provider, original default value is non-empty, while current default value is empty string

**Pull request merged**